src/cunumeric/index/advanced_indexing.cc

```diff
@@ -0,0 +1,109 @@
+/* Copyright 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/index/advanced_indexing.h"
+#include "cunumeric/index/advanced_indexing_template.inl"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <LegateTypeCode CODE, int DIM1, int DIM2, bool IS_SET>
+struct AdvancedIndexingImplBody<VariantKind::CPU, CODE, DIM1, DIM2, IS_SET> {
+  using VAL = legate_type_of<CODE>;
+
+  void compute_output(Buffer<VAL>& out,
+                      const AccessorRO<VAL, DIM1>& input,
+                      const AccessorRO<bool, DIM2>& index,
+                      const Pitches<DIM1 - 1>& pitches_input,
+                      const Rect<DIM1>& rect_input,
+                      const Pitches<DIM2 - 1>& pitches_index,
+                      const Rect<DIM2>& rect_index,
+                      int volume) const
+  {
+    int64_t out_idx = 0;
+    for (size_t idx = 0; idx < volume; ++idx) {
+      auto p       = pitches_index.unflatten(idx, rect_index.lo);
+      auto p_input = pitches_input.unflatten(idx, rect_input.lo);
+      if (index[p] == true) {
+        out[out_idx] = input[p_input];
```

Comment:
I think you can special-case the situation where `DIM1 == DIM2`, and in that case only compute `p` once (since then `rect_index` must be equal to `rect_input`) and use for indexing bth `index[p]` and `input[p]`. And also for that scenario you can emit special code for the "dense" case.

Reply:
@manopapad: I am doing that in the follow-up PR where this task is outputting  ND output regions. Would it be OK for you if I don't change this in this PR?

Reply:
Yes, good point, this PR is large enough as it is.